01-30-shield-blocking/actors/enemy/Enemy.gd

```diff
@@ -0,0 +1,26 @@
+extends KinematicBody2D
+class_name Enemy
+
+onready var animated_sprite : AnimatedSprite = $AnimatedSprite
+
+export var move_speed : = 75.0
+export var attack_distance : = 0
+
+var player
+
+
+func _ready() -> void:
+	set_physics_process(false)
+
+
+func _physics_process(delta: float) -> void:
+	var direction = 1 if player.global_position.x > global_position.x else -1
+	move_and_slide(move_speed * Vector2.RIGHT * direction, Vector2.UP)
+	animated_sprite.flip_h = direction
+	if attack_distance != 0 and global_position.distance_to(player.global_position) <= attack_distance:
```

Comment:
You shouldn't have to check that the distance is not 0, but give it a default value instead. And if you want to assert that the distance isn't negative, it should be in _init, not on every tick in process. Just check for the distance to the player

Reply:
The enemy can work by either attacking the player from code or from contact. If the `attack_distance` is different than 0 this means the enemy will attack from code and not only from a collision. I had a `bool` before but this would just be redundant. 
Another option is to create a new enemy that only attacks from a certain distance and another one which is only by a collision.

Reply:
Okay I didn't notice that. I think you should exclusively rely on collisions for attacks, work with areas interacting with one another. This way you can rely on the well-tested and high-performance backend and simplify the code (in general, in a regular game project thatĺl get bigger over time)